pkg/go/graph/weighted_graph.go

```diff
@@ -0,0 +1,409 @@
+package graph
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+const Infinite = math.MaxInt32
+
+var errModelCycle = fmt.Errorf("model cycle")
+var errTupleCycle = fmt.Errorf("operands AND or BUT NOT cannot be involved in a cycle")
+
+type WeightedAuthorizationModelGraph struct {
+	nodes map[string]*WeightedAuthorizationModelNode
+	edges map[string][]*WeightedAuthorizationModelEdge
```

Comment:
Why does this graph not include `*multi.DirectedGraph` ? By doing this, you are not taking advantage at all of `gonum`. Eg we won't be able to use any of the algorithms on https://github.com/gonum/gonum/tree/master/graph/path. We won't be able to draw the DOT of the model which would help to review the test that you wrote.

Reply:
I didn't use the gonum graph because of the unnecessary complexity it brings to iterate over it, I prefer to use a simple adjacency list instead. Any enhancement to the graph could be easily added like double-linked edges to be used for check or listObjects. The gonum library is good for visualizing the graph, and for that, there is already the AuthorizationModelGraph. This graph is built using the AuthorizationModelGraph, taking all the nodes and edges and building the WeightedAuthorizationModelGraph. The parsing of the model is done through the AuthorizationModelGraphBuilder